app/src/main/java/com/wearerommies/roomie/presentation/navigator/MainNavigator.kt

```diff
@@ -0,0 +1,72 @@
+package com.wearerommies.roomie.presentation.navigator
+
+import android.annotation.SuppressLint
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.navigation.NavDestination
+import androidx.navigation.NavHostController
+import androidx.navigation.compose.currentBackStackEntryAsState
+import androidx.navigation.compose.rememberNavController
+import androidx.navigation.navOptions
+import com.wearerommies.roomie.presentation.model.MainTabRoute
+import com.wearerommies.roomie.presentation.model.Route
+import com.wearerommies.roomie.presentation.type.MainTabType
+import com.wearerommies.roomie.presentation.ui.home.navigateHome
+import com.wearerommies.roomie.presentation.ui.map.navigateMap
+import com.wearerommies.roomie.presentation.ui.mypage.navigateMy
+
+class MainNavigator(
+    val navController: NavHostController,
+) {
+    private val currentDestination: NavDestination?
+        @Composable get() = navController
+            .currentBackStackEntryAsState().value?.destination
+
+    val startDestination = MainTabRoute.Home
+
+    val currentTab: MainTabType?
+        @SuppressLint("RestrictedApi") @Composable get() = MainTabType.find { tab ->
+            currentDestination?.route == tab::class.qualifiedName
+        }
+
+    fun navigate(tab: MainTabType) {
+        val navOptions = navOptions {
+            popUpTo(MainTabType.HOME.route) {
+                saveState = true
+            }
+            launchSingleTop = true
+            restoreState = true
+        }
+
+        when (tab) {
+            MainTabType.HOME -> navController.navigateHome(navOptions)
+            MainTabType.MAP -> navController.navigateMap(navOptions)
+            MainTabType.MY -> navController.navigateMy(navOptions)
+        }
+    }
+
+    private fun popBackStack() {
+        navController.popBackStack()
+    }
```

Comment:
```suggestion
    private fun navigateUp() {
        navController.navigateUp()
    }
```

popBackStack 의 경우에는 현재 백스택을 인식하지 못하게 되어 사용자가 빠르게 더블클릭했을 때 스택 두개가 모두 사라지게 됩니다! 그러므로 이런 예외상황에서도 백스택을 인식하는 navigateUp() 을 사용해주는 것이 좋습니다 :) 아래에 있는 popBackStackIfNotHome 도 동일하게요!